### PR TITLE
Ztest userspace

### LINKED ISSUE #44108: `CONFIG_ZTEST_NEW_API=y` broken with `CONFIG_TEST_USERSPACE=y`
**Describe the bug**
The new ztest API (`CONFIG_ZTEST_NEW_API=y`) is broken when used in combination with `CONFIG_TEST_USERSPACE=y`. There are no in-tree tests of this combination.

One some platforms (e.g. `qemu_cortex_m3`) the test added below fails the assert, indicating the test case is not running in user space after all. On other platforms (e.g. `frdm_k64f`), the test below causes a bus fault.

**To Reproduce**

Apply the following patch to create a test case:

```
diff --git a/tests/ztest/base/prj_verbose_0.conf b/tests/ztest/base/prj_verbose_0.conf
index b4903cd10c..7233d69b58 100644
--- a/tests/ztest/base/prj_verbose_0.conf
+++ b/tests/ztest/base/prj_verbose_0.conf
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_ZTEST_NEW_API=y
 CONFIG_ZTEST_ASSERT_VERBOSE=0
+CONFIG_TEST_USERSPACE=y
diff --git a/tests/ztest/base/src/main.c b/tests/ztest/base/src/main.c
index 07fc9eb474..21d7c0e14d 100644
--- a/tests/ztest/base/src/main.c
+++ b/tests/ztest/base/src/main.c
@@ -35,6 +35,11 @@ ZTEST(framework_tests, test_assert_mem_equal)
        zassert_mem_equal(actual, expected, sizeof(expected), NULL);
 }
 
+ZTEST_USER(framework_tests, test_userspace_is_user)
+{
+       zassert_true(k_is_user_context(), NULL);
+}
+
 /***************************************************************************************************
  * Sample fixture tests
  **************************************************************************************************/
```

1. `west build -b qemu_cortex_m3 tests/ztest/base -- -DCONF_FILE=prj_verbose_0.conf`
2. See error:

```
qemu-system-arm: warning: nic stellaris_enet.0 has no peer
Timer with period zero, disabling
*** Booting Zephyr OS build zephyr-v3.0.0-1639-ga4d438444771  ***
Running test suite fixture_tests
===================================================================
START - test_fixture_pointer
 PASS - test_fixture_pointer in 0.1 seconds
===================================================================
Test suite fixture_tests succeeded
Running test suite framework_tests
===================================================================
START - test_assert_mem_equal
 PASS - test_assert_mem_equal in 0.1 seconds
===================================================================
START - test_assert_tests
 PASS - test_assert_tests in 0.1 seconds
===================================================================
START - test_empty_test
 PASS - test_empty_test in 0.1 seconds
===================================================================
START - test_userspace_is_user

    Assertion failed at WEST_TOPDIR/zephyr/tests/ztest/base/src/main.c:40
 FAIL - test_userspace_is_user in 0.1 seconds
===================================================================
Test suite framework_tests failed.
Running test suite rules_tests
===================================================================
START - test_rules_before_after
 PASS - test_rules_before_after in 0.1 seconds
===================================================================
Test suite rules_tests succeeded
===================================================================
PROJECT EXECUTION FAILED
```

1. `west build -t flash -b frdm_k64f tests/ztest/base -- -DCONF_FILE=prj_verbose_0.conf`
2. See error:3. 

```
*** Booting Zephyr OS build zephyr-v3.0.0-1639-ga4d438444771  ***
Running test suite fixture_tests
===================================================================
START - test_fixture_pointer
 PASS - test_fixture_pointer in 0.1 seconds
===================================================================
Test suite fixture_tests succeeded
Running test suite framework_tests
===================================================================
START - test_assert_mem_equal
 PASS - test_assert_mem_equal in 0.1 seconds
===================================================================
START - test_assert_tests
 PASS - test_assert_tests in 0.1 seconds
===================================================================
START - test_empty_test
 PASS - test_empty_test in 0.1 seconds
===================================================================
START - test_userspace_is_user
E: ***** BUS FAULT *****
E:   Precise data bus error
E:   BFAR Address: 0x20001f6c
E:   NXP MPU error, port 3
E:     Mode: User, Data Address: 0x20001f6c
E:     Type: Read, Master: 0, Regions: 0x8100
E: r0/a1:  0x20001f64  r1/a2:  0x20001fec  r2/a3:  0x00000000
E: r3/a4:  0x00000001 r12/ip:  0x0000ae43 r14/lr:  0x0000ae4f
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x00002c76
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20000020 (test_userspace_is_user)
E: Halting system
```

```
$ addr2line -e build/zephyr/zephyr.elf 0x00002c76
zephyr/subsys/testsuite/ztest/src/ztest_new.c:403
```

**Expected behavior**
`CONFIG_ZTEST_NEW_API=y` should work with `CONFIG_TEST_USERSPACE=y`.

**Environment (please complete the following information):**
 - OS: Ubuntu Linux 20.04 LTS
 - Toolchain: Zephyr SDK v0.13.2
 - Commit SHA: a4d4384447710e9885c0248385db0d5a5d197d50

### PR DESCRIPTION
**Pull request by yperess**
A second attempt at getting userspace to work with ztest's new API. It looks like the addition of the sections to the linker was still breaking the MMU/MPU. This attempt uses the entry point as a `__syscall` which should work around that issue.

Fixes #44108

### CODE REVIEW

**Comment by yperess** [comment]
Running the full tests via `$ ./scripts/twister -v` I'm seeing 82 failures, but then looking at the log I'm only seeing 1:
```
$ cat twister-out/twister.log | grep "FAILED"
2022-04-29 21:54:10,677 - twister - INFO - 20041/35977 lpcxpresso55s69_cpu0      samples/sensor/fxos8700/sample.sensor.fxos8700.accel FAILED Cmake build failure (build)
```

**Comment by yperess** [comment]
@asemjonovs can you take a look at this PR, I believe there's an error in the shuffle logic when I'm moving over to the array of pointers method (which still isn't working because I haven't figured out how to statically allocate that array)

**Review by asemjonovs — CHANGES_REQUESTED** [review]

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 1]
```diff
@@ -598,31 +602,45 @@ void end_report(void)
...
-			sizeof(struct ztest_suite_node));
-#endif
-
+	// DO NOT MERGE, it seems that in C I can't get the number of suites at compile time,
+	// will need another approach
+	static struct ztest_suite_node *suites_to_run[512];
+
+	__ASSERT_NO_MSG(ZTEST_SUITE_COUNT <= 512);
+	z_ztest_shuffle((void **)suites_to_run, (intptr_t)_ztest_suite_node_list_start,
+			ZTEST_SUITE_COUNT, sizeof(struct ztest_suite_node));
```
[the remark this answers is not in the document]
Comment (yperess) [reply subsys/testsuite/ztest/src/ztest_new.c:617]: The size is used to increment the `intptr_t`

**Comment by yperess** [comment]
@dcpleung @andyross I believe that this PR is ready for review now.

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 2]
```diff
@@ -598,26 +595,28 @@ void end_report(void)
...
+	memset(suites_to_run, 0, ZTEST_SUITE_COUNT * sizeof(struct ztest_suite_node *));
+	z_ztest_shuffle((void **)suites_to_run, (intptr_t)_ztest_suite_node_list_start,
+			ZTEST_SUITE_COUNT, sizeof(struct ztest_suite_node));
+	for (size_t i = 0; i < ZTEST_SUITE_COUNT; ++i) {
+		ptr = suites_to_run[i];
+#else
 	for (ptr = _ztest_suite_node_list_start; ptr < _ztest_suite_node_list_end; ++ptr) {
-		struct ztest_suite_stats *stats = &ptr->stats;
+#endif
+		struct ztest_suite_stats *stats = ptr->stats;
```
Comment (asemjonovs) [inline subsys/testsuite/ztest/src/ztest_new.c:614]: NIT, I'm not a fan of the for loop being split by the conditional `ZTEST_SHUFFLE`. Could we make `suites_to_run` work outside of the ifdefs?

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 3]
```diff
@@ -555,12 +543,21 @@ static int z_ztest_run_test_suite_ptr(struct ztest_suite_node *suite)
...
+		z_ztest_shuffle((void **)tests_to_run, (intptr_t)_ztest_unit_test_list_start,
+				ZTEST_TEST_COUNT, sizeof(struct ztest_unit_test));
+		for (size_t i = 0; i < ZTEST_TEST_COUNT; ++i) {
+			test = tests_to_run[i];
+			/* Make sure that the test belongs to this suite */
+			if (strcmp(suite->name, test->test_suite_name) != 0) {
+				continue;
+			}
+#else
 		while (((test = ztest_get_next_test(suite->name, test)) != NULL)) {
```
Comment (asemjonovs) [inline subsys/testsuite/ztest/src/ztest_new.c:559]: We can update ztest_get_next_test to accept the head pointer of the list, so we don't duplicate code here.

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 4]
```diff
@@ -181,11 +176,10 @@ void z_impl_z_test_1cpu_start(void)
...
 	/* coverity[DEADCODE] */
-	for (int i = 0; i < NUM_CPUHOLD; i++)  {
-		k_thread_create(&cpuhold_threads[i],
-				cpuhold_stacks[i], CPUHOLD_STACK_SZ,
-				(k_thread_entry_t) cpu_hold, NULL, NULL, NULL,
-				K_HIGHEST_THREAD_PRIO, 0, K_NO_WAIT);
+	for (int i = 0; i < NUM_CPUHOLD; i++) {
+		k_thread_create(&cpuhold_threads[i], cpuhold_stacks[i], CPUHOLD_STACK_SZ,
+				(k_thread_entry_t)cpu_hold, NULL, NULL, NULL, K_HIGHEST_THREAD_PRIO,
+				0, K_NO_WAIT);
```
Comment (andyross) [inline subsys/testsuite/ztest/src/ztest_new.c:182]: This reminds me that #43609 hasn't merged yet, which is going to collide with the whitespace changes here and require a rebase.  Not a big deal I guess.

**Review by andyross — DISMISSED** [review]
Seems reasonable.  I'll let others nitpick over the ztest internals, but this looks great to me.

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 5]
```diff
@@ -507,20 +501,18 @@ struct ztest_unit_test *ztest_get_next_test(const char *suite, struct ztest_unit
...
-		int j = sys_rand32_get() % (i + 1);
-
-		if (i != j) {
-			memcpy(tmp, arr + (j * elem_size), elem_size);
-			memcpy(arr + (j * elem_size), arr + (i * elem_size), elem_size);
-			memcpy(arr + (i * elem_size), tmp, elem_size);
+		/* Get the next valid position */
+		while (dest[pos] != NULL) {
+			pos = (pos + 1) % num_items;
 		}
```
[the remark this answers is not in the document]
Comment (yperess) [reply subsys/testsuite/ztest/src/ztest_new.c:512]: Added an assert

File: subsys/testsuite/ztest/src/ztest_new.c (hunk 3 above)
Comment (yperess) [reply subsys/testsuite/ztest/src/ztest_new.c:559]: The problem is that we have two different lists, `struct ztest_unit_test *[]` and `struct ztest_unit_test *` so I'm not sure there's a good way for it. I split them up so the whole loop body is in a single `#if`, should make it easier to understand

File: include/zephyr/linker/common-rom/common-rom-kernel-devices.ld [hunk 6]
```diff
@@ -32,6 +32,25 @@
 		__device_end = .;
 	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 
+#if defined(CONFIG_ZTEST)
```
[the remark this answers is not in the document]
Comment (yperess) [reply include/zephyr/linker/common-rom/common-rom-kernel-devices.ld:35]: You're right, I just forgot about it, swapped over to use `Z_LINK_ITERABLE` also removed the `#ifdef`

File: include/zephyr/linker/common-rom/common-rom-kernel-devices.ld [hunk 7]
```diff
@@ -32,6 +32,15 @@
...
 
+	SECTION_PROLOGUE(ztest,,)
+	{
+		Z_LINK_ITERABLE(ztest_suite_node);
+
+		Z_LINK_ITERABLE(ztest_unit_test);
+
+		Z_LINK_ITERABLE(ztest_test_rule);
+	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+
```
[the remark this answers is not in the document]
Comment (yperess) [reply include/zephyr/linker/common-rom/common-rom-kernel-devices.ld]: Done

**Review by andyross — DISMISSED** [review]
Still +1 from me

File: include/zephyr/linker/common-rom/common-rom-kernel-devices.ld (hunk 7 above)
[the remark this answers is not in the document]
Comment (yperess) [reply include/zephyr/linker/common-rom/common-rom-kernel-devices.ld]: @dcpleung I moved things but now it doesn't seem to be including `common-rom-ztest.ld`. I added:
```
#include "common-rom/common-rom-ztest.ld"
```

to `common-rom.ld` which seems to follow the rest of the includes.

[the remark this answers is not in the document]
Comment (yperess) [reply include/zephyr/linker/common-rom/common-rom-kernel-devices.ld]: Ah, the issue is that `soc/*/esp32*/linker.ld` are including the separate components. It seems that the only one not being included is `<zephyr/linker/common-rom/common-rom-logging.ld>`. Should we just change those to all include `<zephyr/linker/common-rom/common-rom.ld>`?

**Comment by yperess** [comment]
@andyross I've made the fix to for out the ztest sections in the common-rom, but the build was broken for esp32 based soc. It seems they do something strange in the linker, anyway it's all working now. Sorry for needing another review.

**Review by andyross — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
Fixed wrong assignee. @yperess please fix the compliance issue.

**Comment by dcpleung** [comment]
The compliance issues are errors about function parameters ending in underscore (being passed to UTIL_CAT()). So I think they are fine as this is how one uses UTIL_CAT().

**Comment by yperess** [comment]
> Fixed wrong assignee. @yperess please fix the compliance issue.

Trying to get clarity on that. I'm seeing 2 issues and both seem incorrect to me:

1. Complaint about newline at end of file. I thought we always want newlines to make the diffs easier and reduce merge conflicts.
2. Complaint about macro arguments needing variable names. The tool thinks that the macro arguments are types and need a name. This is clearly incorrect.

Should I fix just \# 1 then?

**Comment by mbolivar-nordic** [comment]
> * Complaint about newline at end of file. I thought we always want newlines to make the diffs easier and reduce merge conflicts.

This is a complaint about a blank line at end of file (`\n\n`), not a newline at end of file (`\n`).

> * Complaint about macro arguments needing variable names. The tool thinks that the macro arguments are types and need a name. This is clearly incorrect.

Agreed.

> Should I fix just # 1 then?

Yes, but then we'll need someone who can force merge to override compliance, I think.

**Comment by yperess** [comment]
OK, removed the empty line from the `kscan_api/prj.conf`. Need reviews again. Sorry

**Review by mbolivar-nordic — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by stephanosio** [comment]
Merging as per @mbolivar-nordic's request

File: subsys/testsuite/include/ztest.ld [hunk 8]
```diff
@@ -3,26 +3,3 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
-
```
[the remark this answers is not in the document]
Comment (yperess) [reply subsys/testsuite/include/ztest.ld:6]: Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/45987

File: subsys/testsuite/ztest/include/ztest_test_new.h [hunk 9]
```diff
@@ -49,45 +56,47 @@ struct ztest_suite_stats {
  */
 struct ztest_suite_node {
 	/** The name of the test suite. */
-	const char *name;
+	const char * const name;
```
[the remark this answers is not in the document]
Comment (yperess) [reply subsys/testsuite/ztest/include/ztest_test_new.h:59]: Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/45987

File: subsys/testsuite/ztest/include/ztest_test_new.h [hunk 10]
```diff
@@ -49,45 +56,47 @@ struct ztest_suite_stats {
...
 	 * test will only run once on the first attempt.
 	 *
 	 * @param state The current state of the test application.
 	 * @return True if the suite should be run; false to skip.
 	 */
-	bool (*predicate)(const void *state);
+	bool (*const predicate)(const void *state);
 	/** Stats */
-	struct ztest_suite_stats stats;
+	struct ztest_suite_stats * const stats;
```
[the remark this answers is not in the document]
Comment (yperess) [reply subsys/testsuite/ztest/include/ztest_test_new.h:93]: Resolved in https://github.com/zephyrproject-rtos/zephyr/pull/45987

**Pull request merged**